src/ZODB/ConflictResolution.py

```diff
@@ -196,6 +196,9 @@ def __le__(self, other):
     def __ge__(self, other):
         return self.__cmp__(other) >= 0
 
+    def __hash__(self):
+        return hash((self.oid, self.database_name, self.weak))
```

Comment:
I'm not convinced using PersistentReferences as keys in dicts/sets will work correctly when `__eq__` either returns True or raises a ValueError, but never returns False.

I guess I'll see if the tests will manage to convince me.

Reply:
Well. `ValueError` during a `_p_resolveConflict` end in a `ConflictError`, so I suppose I can help in some cases (when comparison is True), but it cannot be worse that currently.

Reply:
@azmeuk I think @mgedmin is correct here:  without reliable equality semantics, hashability cannot be useful.

Reply:
Well I find it useful and I would love this to come into ZODB so I could use it in my projects :)
@tseaver would you agree with those points?

- There is no backward compatibility;
- There is no danger for existing data;
- It might help in some cases;
- It might have unknown behavior in some other cases.

My suggestion is to merge, and let the users open tickets if they find unwanted behavior, as mgedmin suggested it.

Reply:
I think I'm with @tseaver and @mgedmin here and it's probably not a good idea for ZODB to introduce an attractive footgun into conflict resolution.

The claim is that it's "not possible to compare different weak references", so there's not a total order defined and you can't put these in a BTree based Set either.  But you could violate that if you chose, either by making an artificial key to wrap the reference, or by storing the references in a dictionary keyed on their components. You'd be comparing different references, which is apparently not generally reliable, but perhaps in your special circumstance that's good enough?


```python
s = BTrees.OOBTree.TreeSet()
s.add((ref.oid, ref.database_name, ref.weak, ref))

d = {}
d[(ref.oid, ref.database_name, ref.weak)] = ref
```